src/java/htsjdk/samtools/cram/build/CompressionHeaderFactory.java

```diff
@@ -376,429 +337,218 @@ public int compare(final byte[] o1, final byte[] o2) {
-                }
-
-            }
-
-        }
-
-        public Encoding<Integer> getBestEncoding() {
-            if (dictionary != null && dictionary.size() == 1) {
-                final int value = dictionary.keySet().iterator().next();
-                final EncodingParams param = HuffmanIntegerEncoding.toParam(new int[]{value}, new int[]{0});
-                final HuffmanIntegerEncoding huffmanEncoding = new HuffmanIntegerEncoding();
-                huffmanEncoding.fromByteArray(param.params);
-                return huffmanEncoding;
-            }
-
-            EncodingLengthCalculator bestCalculator = calculators.get(0);
-
-            for (final EncodingLengthCalculator calculator : calculators) {
-                if (calculator.length() < bestCalculator.length())
-                    bestCalculator = calculator;
-            }
-
-            Encoding<Integer> bestEncoding = bestCalculator.encoding;
-            long bits = bestCalculator.length();
-
-            { // check if beta is better:
-
-                final int betaLength = (int) Math.round(Math.log(max - minValue) / Math.log(2) + 0.5);
-                if (bits > betaLength * count) {
-                    bestEncoding = new BetaIntegerEncoding(-minValue, betaLength);
-                    bits = betaLength * count;
-                }
-            }
-
-            { // try huffman:
-                if (dictionary != null) {
-                    final HuffmanParamsCalculator huffmanParamsCalculator = new HuffmanParamsCalculator();
-                    for (final Integer value : dictionary.keySet())
-                        huffmanParamsCalculator.add(value, dictionary.get(value).value);
-
-                    huffmanParamsCalculator.calculate();
-
-                    final EncodingParams param = HuffmanIntegerEncoding.toParam(huffmanParamsCalculator.values(), huffmanParamsCalculator.bitLens());
-                    final HuffmanIntegerEncoding huffmanEncoding = new HuffmanIntegerEncoding();
-                    huffmanEncoding.fromByteArray(param.params);
-                    final EncodingLengthCalculator calculator = new EncodingLengthCalculator(huffmanEncoding);
-                    for (final Integer key : dictionary.keySet())
-                        calculator.add(key, dictionary.get(key).value);
-
-                    if (calculator.length() < bits) {
-                        bestEncoding = huffmanEncoding;
-                        bits = calculator.length();
-                    }
-                }
-            }
-
-            byte[] params = bestEncoding.toByteArray();
-            params = Arrays.copyOf(params, Math.min(params.length, 20));
-            log.debug("Best encoding for " + name + ": " + bestEncoding.id().name() + Arrays.toString(params) + ", bits=" + bits);
+    private static class EncodingDetails {
+        ExternalCompressor compressor;
+        EncodingParams params;
 
-            return bestEncoding;
+        public EncodingDetails() {
```

Comment:
remove this no parameter constructor and replace it with a constructor that takes the 2 field values.

Only construct one once you have the correct values in place.